rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -111,39 +117,26 @@ object Substitute {
       case ENeq(par1, par2)   => ENeq(substitute(par1), substitute(par2))
       case EAnd(par1, par2)   => EAnd(substitute(par1), substitute(par2))
       case EOr(par1, par2)    => EOr(substitute(par1), substitute(par2))
-      case EList(ps, _) =>
+      case EList(ps, freeCount, locallyFree) =>
         val _ps = for { par <- ps } yield {
           substitute(par)
         }
-        val fc = (0 /: _ps) { (i, par) =>
-          i + par.freeCount
-        }
-        EList(_ps, fc)
-      case ETuple(ps, _) =>
+        EList(_ps, freeCount, locallyFree.from(env.level).map(x => x - env.level))
+      case ETuple(ps, freeCount, locallyFree) =>
         val _ps = for { par <- ps } yield {
           substitute(par)
         }
-        val fc = (0 /: _ps) { (i, par) =>
-          i + par.freeCount
-        }
-        ETuple(_ps, fc)
-      case ESet(ps, _) =>
+        ETuple(_ps, freeCount, locallyFree.from(env.level).map(x => x - env.level))
+      case ESet(ps, freeCount, locallyFree) =>
         val _ps = for { par <- ps } yield {
           substitute(par)
         }
-        val fc = (0 /: _ps) { (i, par) =>
-          i + par.freeCount
-        }
-        ESet(_ps, fc)
-      case EMap(kvs, _) =>
+        ESet(_ps, freeCount, locallyFree.from(env.level).map(x => x - env.level))
+      case EMap(kvs, freeCount, locallyFree) =>
         val _ps = for { (p1, p2) <- kvs } yield {
           (substitute(p1), substitute(p2))
         }
-        val fc = (0 /: _ps) {
-          case (i, (p1, p2)) =>
-            i + p1.freeCount + p2.freeCount
-        }
-        EMap(_ps, fc)
+        EMap(_ps, freeCount, locallyFree)
```

Comment:
Do we not need a similar pattern here `locallyFree.from(env.level).map(x => x - env.level)`?

Reply:
See related comment below for potential change to this pattern.

Reply:
Good catch.